qiskit/transpiler/passes/synthesis/high_level_synthesis.py

```diff
@@ -418,151 +449,215 @@ def run(self, dag: DAGCircuit) -> DAGCircuit:
-        avoid Python-space `Operation` creation from a `DAGOpNode` if we wouldn't do anything to the
-        node (which is _most_ nodes)."""
-        return (
-            # The fast path is just for Rust-space standard gates (which excludes
-            # `AnnotatedOperation`).
-            node.is_standard_gate()
-            # If it's a controlled gate, we might choose to do funny things to it.
-            and not node.is_controlled_gate()
-            # If there are plugins to try, they need to be tried.
-            and not self._methods_to_try(node.name)
-            # If all the above constraints hold, and it's already supported or the basis translator
-            # can handle it, we'll leave it be.
-            and (
-                self._instruction_supported(node.name, qubits)
-                # This uses unfortunately private details of `EquivalenceLibrary`, but so does the
-                # `BasisTranslator`, and this is supposed to just be temporary til this is moved
-                # into Rust space.
-                or (
-                    self._equiv_lib is not None
-                    and equivalence.Key(name=node.name, num_qubits=node.num_qubits)
-                    in self._equiv_lib._key_to_node_index
-                )
+    def _synthesize_operation(
+        self,
+        operation: Operation,
+        qubits: tuple[int],
+        tracker: QubitTracker,
+    ) -> QuantumCircuit | Operation | DAGCircuit | None:
+        # Try to synthesize the operation. We'll go through the following options:
+        #  (1) Annotations: if the operator is annotated, synthesize the base operation
+        #       and then apply the modifiers.
+        #  (2) High-level objects: try running the battery of high-level synthesis plugins
+        #  (3) Unrolling custom definitions: try defining the operation if it is not yet
+        #       in the set of supported instructions
+        # If any of the above were triggered, we will recurse and go again through these steps
+        # until no further change occurred. If there was no change, we just return ``None``.
+        synthesized = None
+
+        # Try synthesizing via AnnotatedOperation. This is faster than an isinstance check
+        # but a bit less safe since someone could create operations with a ``modifiers`` attribute.
+        if len(modifiers := getattr(operation, "modifiers", [])) > 0:
+            # The base operation must be synthesized without using potential control qubits
+            # used in the modifiers.
+            num_ctrl = sum(
+                mod.num_ctrl_qubits for mod in modifiers if isinstance(mod, ControlModifier)
             )
-        )
+            baseop_qubits = qubits[num_ctrl:]  # reminder: control qubits are the first ones
+            baseop_tracker = tracker.copy()
+            baseop_tracker.drop(qubits[:num_ctrl])  # no access to control qubits
```

Comment:
This is a very minor point. Would it make sense to combine these two lines into one, having the `QubitTracker` initializer receive `qubits_to_drop` (empty set by default)?

Reply:
Yeah we can do that 👍🏻